### PR TITLE
Library in LSP

### PR DESCRIPTION
**Pull request by imkiva**
Summary

- Decoupled LibraryCompiler and LibraryOwner (also needed by #226).
- Refactored LSP to only work with the library system.
- Auto-detect library root and load at startup.
- Generate mocked library for the single source file (like rust-playground emacs plugin).
- Go to definition across modules.

Depends on aya-prover/aya-vscode#18

### CODE REVIEW

File: base/src/main/java/org/aya/concrete/parse/AyaParsing.java [hunk 1]
```diff
@@ -76,7 +76,7 @@
     @NotNull Reporter reporter, @NotNull String text,
     @NotNull BiFunction<AyaProducer, AyaParser, T> tree
   ) {
-    var sourceFile = new SourceFile(Path.of("stdin"), text);
+    var sourceFile = new SourceFile("<stdin>", Path.of("stdin"), text);
```
Comment (ice1000) [inline base/src/main/java/org/aya/concrete/parse/AyaParsing.java:79]: Please, don't do this. On Windows, it'll say that it's an invalid path.

File: base/src/main/java/org/aya/concrete/visitor/StmtConsumer.java [hunk 2]
```diff
@@ -31,7 +31,7 @@ default void visitClause(@NotNull Pattern.Clause c, P pp) {
 
   default void visitPattern(@NotNull Pattern pattern, P p) {
     switch (pattern) {
-      case Pattern.Ctor ctor -> ctor.params().forEach(param -> visitPattern(pattern, p));
+      case Pattern.Ctor ctor -> ctor.params().forEach(pat -> visitPattern(pat, p));
```
Comment (ice1000) [inline base/src/main/java/org/aya/concrete/visitor/StmtConsumer.java:34]: Thanks!!!! Lmfao

File: lsp/src/main/java/org/aya/lsp/library/WsLibrary.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * @author kiva
+ */
+public record WsLibrary(
+  @NotNull CountingReporter reporter,
+  @NotNull SourceFileLocator locator,
+  @NotNull DynamicSeq<LibrarySource> sources,
+  @NotNull LibraryConfig mockConfig,
+  @NotNull Path workspace
+) implements LibraryOwner {
+  public static @NotNull WsLibrary from(@NotNull Reporter outReporter, @NotNull Path ayaSource) {
```
Comment (ice1000) [inline lsp/src/main/java/org/aya/lsp/library/WsLibrary.java:31]: I propose renaming to `mock`

File: tools/src/main/java/org/aya/util/error/SourceFile.java [hunk 4]
```diff
@@ -6,23 +6,20 @@
...
+  @NotNull String sourceCode
+) {
+  public SourceFile(@NotNull String display, @NotNull Path underlying, @NotNull String sourceCode) {
+    this(display, Option.some(underlying), sourceCode);
   }
 
-  public static final SourceFile NONE = new SourceFile(Option.none(), "");
-  public static final SourceFile SER = new SourceFile(Option.none(), "");
+  public static final SourceFile NONE = new SourceFile("<unknown-file>", Option.none(), "");
+  public static final SourceFile SER = new SourceFile("<serialized-core>", Option.none(), "");
```
Comment (ice1000) [inline tools/src/main/java/org/aya/util/error/SourceFile.java:20]: Ditto about path names

**Review by ice1000 — COMMENTED** [review]
Does WsLibrary work with single file opened as a project?

**Comment by ice1000** [comment]
> Does WsLibrary work with single file opened as a project?

I think the answer is already in the summary. I was dumb

File: base/src/main/java/org/aya/concrete/parse/AyaParsing.java (hunk 1 above)
Comment (imkiva) [reply base/src/main/java/org/aya/concrete/parse/AyaParsing.java:79]: Not anymore. The display field is just a plain `String` now :)

File: base/src/main/java/org/aya/concrete/visitor/StmtConsumer.java (hunk 2 above)
Comment (imkiva) [reply base/src/main/java/org/aya/concrete/visitor/StmtConsumer.java:34]: Copilot did it. 🤣

File: tools/src/main/java/org/aya/util/error/SourceFile.java (hunk 4 above)
Comment (imkiva) [reply tools/src/main/java/org/aya/util/error/SourceFile.java:20]: They will not be used to construct a `Path`, so it is safe to keep brackets.

Comment (imkiva) [reply tools/src/main/java/org/aya/util/error/SourceFile.java:20]: Instead, we store the real path in `SourceFile#underlying`

File: base/src/main/java/org/aya/concrete/parse/AyaParsing.java (hunk 1 above)
Comment (ice1000) [reply base/src/main/java/org/aya/concrete/parse/AyaParsing.java:79]: Noice.

File: tools/src/main/java/org/aya/util/error/SourceFile.java (hunk 4 above)
Comment (ice1000) [reply tools/src/main/java/org/aya/util/error/SourceFile.java:20]: Noice

**Comment by imkiva** [comment]
> Btw have you resolved the problem about the 'triggering exception in compiler breaks highlighting' problem?

Not yet. Maybe a single try-catch would do

**Pull request merged**